### PR TITLE
NumPy docstring style

### LINKED ISSUE #315: Convert all docstrings to numpy style
As part of our overall docs overhaul (see glotzerlab/signac-docs#64), we want to convert our docstrings to numpy style (as decided in glotzerlab/signac-docs#74). The best automated tool I'm familiar with for this task is [pyment](https://github.com/dadadel/pyment). In addition to converting docstrings, it will also generate docstrings for functions, classes, etc that are missing docstrings entirely. However, the conversion will require significant manual review to ensure that all docstrings are converted correctly.

### PR DESCRIPTION
**Pull request by bdice**
## Description
This PR combines work from many contributors, rewriting API docstrings to adhere to NumPy docstring style.

_We need to determine next steps for this PR. How many reviews / how much detail will suffice, since all work has been previously reviewed in some form? See discussion in #315._

# [ReadTheDocs build of this branch](https://docs.signac.io/projects/core/en/numpy_docs/)

## Motivation and Context
May resolve #315, depending on further assessment of needed changes.

## Types of Changes

- [x] Documentation update

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:
- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt) and added all related issue and pull request numbers for future reference (if applicable). See example below.

### CODE REVIEW

**Comment by vyasr** [comment]
@bdice I assume the removal of the assigned reviewers for this PR was accidental (and probably not done by you at all)? I reassigned since I noticed this PR didn't have any reviewers, but feel free to change if there was a good reason for removal.

**Comment by bdice** [comment]
@vyasr This PR is large and I wanted to have a short strategic discussion during our dev meeting before assigning reviewers. Let's chat about it in 1 hour.

File: signac/contrib/utility.py [hunk 1]
```diff
@@ -274,9 +428,22 @@ def _encode_tree(x):
...
+        By default, values are encoded to be hashable. Use ``None`` to skip encoding.
+    key : str
+        Key of root at current point in the recursion, used to
+        build up nested keys in the top-level dict through
+        multiple recursive calls (Default value = None).
+
+    Yields
+    ------
+    tuple
+        Tupples of dotted key and values e.g. ('a.b', 'c')
```
Comment (atravitz) [inline signac/contrib/utility.py:445]:
```suggestion
        Tuples of dotted key and values e.g. ('a.b', 'c')
```

**Comment by mikemhenry** [comment]
Okay once these tests pass we should be good to merge!

**Comment by bdice** [comment]
🎉 Wow, great team effort! Thanks to everyone who committed and reviewed (too many to list)!

**Pull request merged**